non-breaking change to rolled array

cajahnke opened the pull request:
Hi Troy,

I promise to leave this alone henceforth, but once I started using node-roll in practice I started wishing that rolled would reflect the actual dice rolled on compound dice rolls, eg 2d20b1+d4 would return an array containing the array of the two d20 rolls and an array containing the d4 roll.

Your mileage may vary as to how useful that is, but it matches my intuition about what data we would want from rolled on compound rolls.  Unfortunately, it means current tests for the contents of rolled would need to be rewritten to be tests of rolled[0].

Also changed in the test.js file is a change to paths on the two "bin/roll" tests that is necessary for running the test set on my Windows-based system.  I tried to make that part non-breaking for your original tests but wasn't sure of  your environment to test.

cajahnke commented:
Hi Troy,
Assuming that the last pull request was ignored because it would break old implementations that did something with rolled array.  I updated my proposed change to make the rolled array correspond to what the tests require and any hypothetical implementation would expect.  rolled will still contain an array of arrays in cases where you're rolling something odd like d4 + 2d6 + 7, so that the d4 is easier to distinguish from the 2d6.

troygoode commented:
I like this @cajahnke. Would you mind updating the README to illustrate how this works as part of this PR?

cajahnke commented:
just did, plus one other quick fix.  Do I need another pr or can you merege those in from the original?

cajahnke commented:
Thanks, Troy.  Just updated the readme and realized I had the order of rolled in a slightly counterintuitive first=last order, so I added an unshift/pop to correct for that.  No change to any existing tests, though I didn’t add a specific test for the new function.  Maybe I ought to?

From: Troy Goode [mailto:notifications@github.com]
Sent: Thursday, October 29, 2015 12:11 PM
To: troygoode/node-roll <node-roll@noreply.github.com>
Cc: Jahnke, Christopher A <cajahnke@utexas.edu>
Subject: Re: [node-roll] non-breaking change to rolled array (#6)


I like this @cajahnke<https://github.com/cajahnke>. Would you mind updating the README to illustrate how this works as part of this PR?

—
Reply to this email directly or view it on GitHub<https://github.com/troygoode/node-roll/pull/6#issuecomment-152252089>.

cajahnke commented:
Just a quick note to say I’ve updated the readme, as requested, and added a test for the new function.  Note that the rolls invoked by the parser occur before the main roll even though the rolled array gets reordered, so the expected values for the test aren’t [[5,13],[3]] as you might expect.

Fair warning about the readme update – you pretty much have to know D&D 5e for my example variable name to not seem like a reference to ecclesiastical cross-trainers, feel free to make it less idiosyncratric.

From: Troy Goode [mailto:notifications@github.com]
Sent: Thursday, October 29, 2015 12:11 PM
To: troygoode/node-roll <node-roll@noreply.github.com>
Cc: Jahnke, Christopher A <cajahnke@utexas.edu>
Subject: Re: [node-roll] non-breaking change to rolled array (#6)


I like this @cajahnke<https://github.com/cajahnke>. Would you mind updating the README to illustrate how this works as part of this PR?

—
Reply to this email directly or view it on GitHub<https://github.com/troygoode/node-roll/pull/6#issuecomment-152252089>.

cajahnke commented:
Well, hell.  Are we using different versions of Should or Mocha?  I used deepEqual in the test I added for the new function and it ran fine.  Travis build is failing over the new test.  I could take that out, but it seems like a test is a good thing.  How do you test for array member identity in Mocha if there isn’t deepEqual?  Break it into a test of array length plus a series of individual member tests?

From: Jahnke, Christopher A
Sent: Monday, November 02, 2015 2:40 PM
To: 'troygoode/node-roll' <reply@reply.github.com>; 'troygoode@gmail.com' <troygoode@gmail.com>
Subject: RE: [node-roll] non-breaking change to rolled array (#6)

Just a quick note to say I’ve updated the readme, as requested, and added a test for the new function.  Note that the rolls invoked by the parser occur before the main roll even though the rolled array gets reordered, so the expected values for the test aren’t [[5,13],[3]] as you might expect.

Fair warning about the readme update – you pretty much have to know D&D 5e for my example variable name to not seem like a reference to ecclesiastical cross-trainers, feel free to make it less idiosyncratric.

From: Troy Goode [mailto:notifications@github.com]
Sent: Thursday, October 29, 2015 12:11 PM
To: troygoode/node-roll <node-roll@noreply.github.com<mailto:node-roll@noreply.github.com>>
Cc: Jahnke, Christopher A <cajahnke@utexas.edu<mailto:cajahnke@utexas.edu>>
Subject: Re: [node-roll] non-breaking change to rolled array (#6)


I like this @cajahnke<https://github.com/cajahnke>. Would you mind updating the README to illustrate how this works as part of this PR?

—
Reply to this email directly or view it on GitHub<https://github.com/troygoode/node-roll/pull/6#issuecomment-152252089>.

cajahnke commented:
Nm, I see Should has some quirks (http://stackoverflow.com/questions/19235374/how-does-one-do-a-deepequal-assertion-with-should-js).  Have adjusted the test so that the build passes.

From: Jahnke, Christopher A
Sent: Monday, November 02, 2015 2:47 PM
To: 'troygoode/node-roll' <reply@reply.github.com>; 'troygoode@gmail.com' <troygoode@gmail.com>
Subject: RE: [node-roll] non-breaking change to rolled array (#6)

Well, hell.  Are we using different versions of Should or Mocha?  I used deepEqual in the test I added for the new function and it ran fine.  Travis build is failing over the new test.  I could take that out, but it seems like a test is a good thing.  How do you test for array member identity in Mocha if there isn’t deepEqual?  Break it into a test of array length plus a series of individual member tests?

From: Jahnke, Christopher A
Sent: Monday, November 02, 2015 2:40 PM
To: 'troygoode/node-roll' <reply@reply.github.com>; 'troygoode@gmail.com' <troygoode@gmail.com<mailto:troygoode@gmail.com>>
Subject: RE: [node-roll] non-breaking change to rolled array (#6)

Just a quick note to say I’ve updated the readme, as requested, and added a test for the new function.  Note that the rolls invoked by the parser occur before the main roll even though the rolled array gets reordered, so the expected values for the test aren’t [[5,13],[3]] as you might expect.

Fair warning about the readme update – you pretty much have to know D&D 5e for my example variable name to not seem like a reference to ecclesiastical cross-trainers, feel free to make it less idiosyncratric.

From: Troy Goode [mailto:notifications@github.com]
Sent: Thursday, October 29, 2015 12:11 PM
To: troygoode/node-roll <node-roll@noreply.github.com<mailto:node-roll@noreply.github.com>>
Cc: Jahnke, Christopher A <cajahnke@utexas.edu<mailto:cajahnke@utexas.edu>>
Subject: Re: [node-roll] non-breaking change to rolled array (#6)


I like this @cajahnke<https://github.com/cajahnke>. Would you mind updating the README to illustrate how this works as part of this PR?

—
Reply to this email directly or view it on GitHub<https://github.com/troygoode/node-roll/pull/6#issuecomment-152252089>.

troygoode commented:
FYI I'm traveling abroad right now. Might not be able to look at this until I return next week.

cajahnke commented:
oops.  did not think to follow up writing to you after updating the readme, but I did do thaty a while back.  No changes in the past couple of months, just wanted to see if I could perform a littl necromancy on this pull request

cajahnke commented:
No movement on my end, just bumping the email thread for visibility – please let me know if there are outstanding issues preventing you accepting the PR

From: Jahnke, Christopher A
Sent: Monday, November 02, 2015 2:40 PM
To: 'troygoode/node-roll' <reply@reply.github.com>; 'troygoode@gmail.com' <troygoode@gmail.com>
Subject: RE: [node-roll] non-breaking change to rolled array (#6)

Just a quick note to say I’ve updated the readme, as requested, and added a test for the new function.  Note that the rolls invoked by the parser occur before the main roll even though the rolled array gets reordered, so the expected values for the test aren’t [[5,13],[3]] as you might expect.

Fair warning about the readme update – you pretty much have to know D&D 5e for my example variable name to not seem like a reference to ecclesiastical cross-trainers, feel free to make it less idiosyncratric.

From: Troy Goode [mailto:notifications@github.com]
Sent: Thursday, October 29, 2015 12:11 PM
To: troygoode/node-roll <node-roll@noreply.github.com<mailto:node-roll@noreply.github.com>>
Cc: Jahnke, Christopher A <cajahnke@utexas.edu<mailto:cajahnke@utexas.edu>>
Subject: Re: [node-roll] non-breaking change to rolled array (#6)


I like this @cajahnke<https://github.com/cajahnke>. Would you mind updating the README to illustrate how this works as part of this PR?

—
Reply to this email directly or view it on GitHub<https://github.com/troygoode/node-roll/pull/6#issuecomment-152252089>.

cajahnke commented:
Again, just checking in to see if there’s something unacceptable about the documentation update, test addition, or new function that would stop you from accepting this old pull request – no worries if you don’t want to accept but it would clean up my install scripts that tiny bit if I could pull from the main branch.

From: Jahnke, Christopher A
Sent: Monday, February 15, 2016 1:12 PM
To: 'troygoode/node-roll' <reply@reply.github.com>; 'troygoode@gmail.com' <troygoode@gmail.com>
Subject: RE: [node-roll] non-breaking change to rolled array (#6)

No movement on my end, just bumping the email thread for visibility – please let me know if there are outstanding issues preventing you accepting the PR

From: Jahnke, Christopher A
Sent: Monday, November 02, 2015 2:40 PM
To: 'troygoode/node-roll' <reply@reply.github.com>; 'troygoode@gmail.com' <troygoode@gmail.com<mailto:troygoode@gmail.com>>
Subject: RE: [node-roll] non-breaking change to rolled array (#6)

Just a quick note to say I’ve updated the readme, as requested, and added a test for the new function.  Note that the rolls invoked by the parser occur before the main roll even though the rolled array gets reordered, so the expected values for the test aren’t [[5,13],[3]] as you might expect.

Fair warning about the readme update – you pretty much have to know D&D 5e for my example variable name to not seem like a reference to ecclesiastical cross-trainers, feel free to make it less idiosyncratric.

From: Troy Goode [mailto:notifications@github.com]
Sent: Thursday, October 29, 2015 12:11 PM
To: troygoode/node-roll <node-roll@noreply.github.com<mailto:node-roll@noreply.github.com>>
Cc: Jahnke, Christopher A <cajahnke@utexas.edu<mailto:cajahnke@utexas.edu>>
Subject: Re: [node-roll] non-breaking change to rolled array (#6)


I like this @cajahnke<https://github.com/cajahnke>. Would you mind updating the README to illustrate how this works as part of this PR?

—
Reply to this email directly or view it on GitHub<https://github.com/troygoode/node-roll/pull/6#issuecomment-152252089>.

troygoode commented:
Published to npm as `1.2.0`.

I'm not entirely happy with how some of it turned out, but it has nothing to do with your changes. Fixing this "the right way" is going to be tough and will probably need to be reserved for a `2.0.0` that does actual lexer/parser work on the input string.

Specifically I don't care for `.parse` calling `.roll` – while `.parse` wasn't documented it was intended to be "non-executing" and return a readable description of the input rather than an actual result. That is muddied a bit now, but is unavoidable without substantially rewriting `.roll`.

cajahnke commented:
Do you have a problem with .roll calling .parse to parse the input? Or just the other way around?

troygoode commented:
The original idea is that `.roll` would expect a configuration object that is output via `.parse` and that `.roll` would be smart enough that if it received a string it would automatically call `.parse` on that. Again, not asking you to change anything since, for instance, I think we'd have to majorly change the expected output of `.parse` as well (its output isn't currently capable of modeling multiple sets of dice). Just complaining about my own lack of foresight 5 years ago. :-)

cajahnke commented:
No worries.  Work and side projects make promising anything way reckless right now, but if you don't jump on this right away I'll do it myself after I finish eating the present whale (hard to put it down right now).  Could be a few months, but I have a rough idea how it would go and it strikes me as not that bad.
